File: SculptLayerNode/src/SculptLayerNode.cpp
#include <stack>
#include <maya/MDataHandle.h>
#include <maya/MFloatPoint.h>
#include <maya/MFloatVector.h>
#include <maya/MFnData.h>
#include <maya/MFnMesh.h>
#include <maya/MFnMeshData.h>
#include <maya/MFnNumericAttribute.h>
#include <maya/MFnNumericData.h>
#include <maya/MFnTypedAttribute.h>
#include <maya/MGlobal.h>
#include <maya/MIntArray.h>
#include <maya/MPointArray.h>
#include <maya/MVector.h>
#include <maya/MVectorArray.h>
#include "SculptLayerNode.h"

#include <string>
#include <iostream>

//-----------------------------------------------------------------------------
// Set static members
MTypeId SculptLayer::m_id(0x1005);
const MString SculptLayer::typeName("SculptLayerNode");
MObject SculptLayer::m_terrain;
MObject SculptLayer::m_curveMask;
MObject SculptLayer::m_sculptedMesh;
MObject SculptLayer::m_sculptStrength;
MObject SculptLayer::m_curveOffset;
MObject SculptLayer::m_maxProjectionDistance;
MObject SculptLayer::m_outMesh;
//-----------------------------------------------------------------------------
void* SculptLayer::creator()
{
    return new SculptLayer();
}
//-----------------------------------------------------------------------------
MStatus SculptLayer::initialize()
{
	MStatus stat;
	MFnTypedAttribute typedAttr;
	MFnNumericAttribute numericAttr;

	// Input node attributes
	m_terrain = typedAttr.create("terrain", "t", MFnData::kMesh);
	typedAttr.setReadable(false);
	typedAttr.setWritable(true);
	typedAttr.setStorable(true);
	stat = addAttribute(m_terrain);
	if (!stat)
		return stat;

	m_curveMask = typedAttr.create("curveMask", "cm", MFnData::kNurbsCurve);
	typedAttr.setReadable(false);
	typedAttr.setWritable(true);
	typedAttr.setStorable(true);
	stat = addAttribute(m_curveMask);
	if (!stat)
		return stat;

	m_sculptedMesh = typedAttr.create("sculptedMesh", "sm", MFnData::kMesh);
	typedAttr.setReadable(false);
	typedAttr.setWritable(true);
	typedAttr.setStorable(true);
	stat = addAttribute(m_sculptedMesh);
	if (!stat)
		return stat;

    m_sculptStrength = numericAttr.create("sculptStrength", "ss", MFnNumericData::kFloat, 1.0);
    numericAttr.setReadable(false);
    numericAttr.setWritable(true);
    numericAttr.setStorable(true);
	stat = addAttribute(m_sculptStrength);
	if (!stat)
		return stat;

    m_curveOffset = numericAttr.create("curveOffset", "co", MFnNumericData::kFloat, 1.1);
    numericAttr.setReadable(false);
    numericAttr.setWritable(true);
    numericAttr.setStorable(true);
    stat = addAttribute(m_curveOffset);
    if (!stat)
        return stat;

    m_maxProjectionDistance = numericAttr.create("maxProjectionDistance", "mpd", MFnNumericData::kFloat, 1000);
    numericAttr.setReadable(false);
    numericAttr.setWritable(true);
    numericAttr.setStorable(true);
    stat = addAttribute(m_maxProjectionDistance);
    if (!stat)
        return stat;

	// Output node attributes
	m_outMesh = typedAttr.create("outMesh", "m", MFnData::kMesh);
	typedAttr.setReadable(true);
	typedAttr.setWritable(false);
	typedAttr.setStorable(false);
	stat = addAttribute(m_outMesh);
	if (!stat)
		return stat;

	// Connect input/output dependencies
	attributeAffects(m_terrain, m_outMesh);
	attributeAffects(m_curveMask, m_outMesh);
	attributeAffects(m_sculptedMesh, m_outMesh);
	attributeAffects(m_sculptStrength, m_outMesh);
    attributeAffects(m_curveOffset, m_outMesh);
    attributeAffects(m_maxProjectionDistance, m_outMesh);

	// Return success
	return MStatus::kSuccess;
}
//-----------------------------------------------------------------------------
MStatus SculptLayer::compute(const MPlug &_plug, MDataBlock &_data)
{
	MStatus stat;
	// Check if the plug is the output plug
	if (_plug == m_outMesh)
	{
		// Get the data handles for inputs and convert to the correct type
        MDataHandle terrainDataHandle = _data.inputValue(m_terrain);
		MObject terrainValue = terrainDataHandle.asMesh();

        MDataHandle curveMaskDataHandle = _data.inputValue(m_curveMask);
		MObject curveMaskValue = curveMaskDataHandle.asNurbsCurve();

        MDataHandle sculptedMeshDataHandle = _data.inputValue(m_sculptedMesh);
		MObject sculptedMeshValue = sculptedMeshDataHandle.asMesh();

        MDataHandle sculptStrengthDataHandle = _data.inputValue(m_sculptStrength);
        float sculptStrengthValue = sculptStrengthDataHandle.asFloat();

        MDataHandle curveOffsetDataHandle = _data.inputValue(m_curveOffset);
        float curveOffsetValue = curveOffsetDataHandle.asFloat();

        MDataHandle maxProjectionDataHandle = _data.inputValue(m_maxProjectionDistance);
        float maxProjectionDistanceValue = maxProjectionDataHandle.asFloat();

		// Get the output data handle
        MDataHandle outMeshDataHandle = _data.outputValue(m_outMesh);

        // Computation
        // Get the vertices of the original mesh
        MFnMesh inTerrainFn(terrainValue);
        MPointArray vertices;
        inTerrainFn.getPoints(vertices, MSpace::kWorld);

        MFnNurbsCurve curveFn(curveMaskValue);

        bool recompute = false;
        if (m_firstCompute == true)
        {
            recompute = true;
            curveFn.getCVs(m_lastCurvePoints, MSpace::kWorld);
            m_lastCurveOffset = curveOffsetValue;
            int curveCentreClosestVertex;
            MPoint curveCentre = findCurveCentre(curveMaskValue);
            MPoint tempPoint;
            inTerrainFn.getClosestPoint(curveCentre, tempPoint, MSpace::kWorld, &curveCentreClosestVertex);
            m_curveCentreClosestVertex = curveCentreClosestVertex;
            m_firstCompute = false;
        }
        else if (m_lastCurvePoints.length() != curveFn.numCVs())
        {
            recompute = true;
            curveFn.getCVs(m_lastCurvePoints, MSpace::kWorld);
        }
        else if (m_lastCurveOffset != curveOffsetValue)
        {
            recompute = true;
            m_lastCurveOffset = curveOffsetValue;
        }
        else
        {
            int curveCentreClosestVertex;
            MPoint curveCentre = findCurveCentre(curveMaskValue);
            MPoint tempPoint;
            inTerrainFn.getClosestPoint(curveCentre, tempPoint, MSpace::kWorld, &curveCentreClosestVertex);
            if (m_curveCentreClosestVertex != curveCentreClosestVertex)
            {
                recompute = true;
                m_curveCentreClosestVertex = curveCentreClosestVertex;
            }
        }

        // Only compute data if necessary
        if (recompute == true)
        {
            // Find the vertices that lie within the curve
            findVerticesInsideCurve(terrainValue, curveMaskValue, curveOffsetValue);
            // Calculate the soft select values
            calculateSoftSelectValues(vertices, curveMaskValue);
        }

        // Create a function set for the sculpted mesh and acceleration parameters
        MFnMesh sculptedMeshFn(sculptedMeshValue);
        MMeshIsectAccelParams accelerationParams = sculptedMeshFn.autoUniformGridParams();

        // Iterate through the vertices and project onto the sculpted mesh
        MFloatPoint raySource;
        MVector normal;
        MFloatVector rayDirection;
        MFloatPoint intersectionPoint;
        for (auto vertex : m_affectedVerticesData)
        {
            raySource = MFloatPoint(vertices[vertex.first]);
            inTerrainFn.getVertexNormal(vertex.first, true, normal, MSpace::kWorld);
            rayDirection = MFloatVector(normal);
            sculptedMeshFn.closestIntersection(raySource, rayDirection, NULL, NULL, false, MSpace::kWorld, maxProjectionDistanceValue, true, &accelerationParams, intersectionPoint, NULL, NULL, NULL, NULL, NULL, 1e-6f, NULL);
            MVector displacement = MPoint(intersectionPoint) - vertices[vertex.first];
            // Ensure the vertices are not sliding perpendicular to the normal
            if (displacement * normal != 0.0)
            {
                vertices[vertex.first] += displacement * sculptStrengthValue * vertex.second;
            }
        }

        // Create a copy of the terrain
        MFnMeshData meshDataFn;
        MObject outMeshObj = meshDataFn.create();
        MFnMesh newMeshFn;
        newMeshFn.copy(terrainValue, outMeshObj);

        // Set the output mesh vertices
        newMeshFn.setPoints(vertices);
        newMeshFn.setObject(outMeshObj);

		// Set the output value
        outMeshDataHandle.set(outMeshObj);

		// Set the plug as clean
		stat = _data.setClean(_plug);
		if (!stat)
			return stat;

        return MStatus::kSuccess;
	}
    return MStatus::kUnknownParameter;
}
//-----------------------------------------------------------------------------
MPoint SculptLayer::findCurveCentre(const MObject &_curve)
{
    MFnNurbsCurve curveFn(_curve);
    MPoint point;
    int numPoints = curveFn.numCVs() * 2;
    MVector curveCentre(0.0, 0.0, 0.0);
    for (int i=0; i<numPoints; ++i)
    {
        curveFn.getPointAtParam(float(i)/numPoints, point, MSpace::kWorld);
        curveCentre += MVector(point);
    }
    curveCentre /= float(numPoints);
    return MPoint(curveCentre);
}
//-----------------------------------------------------------------------------
void SculptLayer::findVerticesInsideCurve(const MObject &_terrain, const MObject &_curve, const float &_curveOffset)
{
    m_affectedVertices.clear();
    // Create two sets of checked and unchecked face IDs
    std::stack<int> uncheckedFaces;
    std::unordered_set<int> checkedFaces;
    // Calculate the starting index to check
    MPoint curveCentre = findCurveCentre(_curve);
    MFnMesh terrainFn(_terrain);
    MPoint tempPoint;
    int startID;
    terrainFn.getClosestPoint(curveCentre, tempPoint, MSpace::kWorld, &startID);
    // Append the start ID to the checked and unchecked face sets
    uncheckedFaces.emplace(startID);
    checkedFaces.emplace(startID);

    // Create a polygon iterator
    MItMeshPolygon polygonIterator(_terrain);
    int tempIndex;

    MFnNurbsCurve curveFn(_curve);
    MPoint faceCentre;
    MPoint closestPointOnCurve;
    MVector centreToFace;
    MVector centreToCurve;

    while (!uncheckedFaces.empty())
    {
        // Set the iterator to the top of the stack
        polygonIterator.setIndex(uncheckedFaces.top(), tempIndex);
        // Pop from the stack
        uncheckedFaces.pop();
        // Find the face centre
        faceCentre = polygonIterator.center(MSpace::kWorld);
        closestPointOnCurve = curveFn.closestPoint(faceCentre, NULL, kMFnNurbsEpsilon, MSpace::kWorld);
        centreToFace = (faceCentre - curveCentre) * _curveOffset;
        centreToCurve = closestPointOnCurve - curveCentre;
        // Use the dot product to check if the point is inside
        if (centreToFace * centreToCurve.normal() < centreToCurve.length())
        {
            // Get the vertices of the face and append to the set of affected vertices
            MIntArray polyVertices;
            polygonIterator.getVertices(polyVertices);
            for (unsigned i=0; i<polyVertices.length(); ++i)
            {
                m_affectedVertices.emplace(polyVertices[i]);
            }
            // Find the connected faces and append to the list of unchecked faces
            MIntArray connectedFaces;
            polygonIterator.getConnectedFaces(connectedFaces);
            for (unsigned i=0; i<connectedFaces.length(); ++i)
            {
                // If it is not in checked faces, add to the stack
                auto search = checkedFaces.find(connectedFaces[i]);
                if (search == checkedFaces.end())
                {
                    uncheckedFaces.emplace(connectedFaces[i]);
                }
                // Add to the checked faces
                checkedFaces.emplace(connectedFaces[i]);
            }
        }
    }
}
//-----------------------------------------------------------------------------
void SculptLayer::calculateSoftSelectValues(const MPointArray &_vertices, const MObject &_curve)
{
    // Clear the soft select values
    m_affectedVerticesData.clear();

    // Curve centre
    MPoint curveCentre = findCurveCentre(_curve);
    MFnNurbsCurve curveFn(_curve);

    // Closest point on curve
    MPoint curvePoint;
    float centreVertexLenSquared;
    float centreCurvePointLenSquared;
    float ratio;
    float softSelectValue;

    for (int vertexID : m_affectedVertices)
    {
        // Find the closest point on the curve
        curvePoint = curveFn.closestPoint(_vertices[vertexID], NULL, kMFnNurbsEpsilon, MSpace::kWorld);
        centreVertexLenSquared = ((curveCentre.x - _vertices[vertexID].x) * (curveCentre.x - _vertices[vertexID].x)) + ((curveCentre.z - _vertices[vertexID].z) * (curveCentre.z - _vertices[vertexID].z));
        centreCurvePointLenSquared = ((curveCentre.x - curvePoint.x) * (curveCentre.x - curvePoint.x)) + ((curveCentre.z - curvePoint.z) * (curveCentre.z - curvePoint.z));
        ratio = centreVertexLenSquared / centreCurvePointLenSquared;
        softSelectValue = 2.0 + (2.0/(ratio - 2.0));
        if (softSelectValue < 0)
        {
            softSelectValue = 0.0;
        }
        m_affectedVerticesData.emplace_back(std::make_pair(vertexID, softSelectValue));
    }
}
//-----------------------------------------------------------------------------
SculptLayer::SculptLayer(){}
//-----------------------------------------------------------------------------
SculptLayer::~SculptLayer(){}
//-----------------------------------------------------------------------------

File: plugin/SculptLayerNode.py
## SculptLayerNode.py
# This node creats a sculpt layer

import sys
import maya.api.OpenMaya as om

#----------------------------------------------------------
# Plugin
#----------------------------------------------------------

# Node info
kPluginNodeName = "SculptLayerNode"
kPluginNodeID = om.MTypeId(0x1005)

class SculptNodeClass(om.MPxNode):
	# Define the attributes
	# Input objects
	m_terrain = om.MObject()
	m_curveMask = om.MObject()
	m_sculptedMesh = om.MObject()
	# Parameters
	m_sculptStrength = om.MObject()
	m_curveOffset = om.MObject()
	m_maxProjectionDistance = om.MObject()
	# Output
	m_outMesh = om.MObject()

	def __init__(self):
		om.MPxNode.__init__(self)
		self.m_curveOriginalPoints = None
		self.m_affectedVertices = []
		self.m_lastCurveOffset = 0.0
		self.m_lastNumVertices = 0
		self.m_lastVertex01 = []

	## The computation of the node
	def compute(self, _plug, _dataBlock):
		# Check if the plug is the output
		if (_plug == SculptNodeClass.m_outMesh):

			# Get data handles and typecast
			terrainDataHandle = _dataBlock.inputValue(SculptNodeClass.m_terrain)
			terrainValue = terrainDataHandle.asMesh()

			curveMaskDataHandle = _dataBlock.inputValue(SculptNodeClass.m_curveMask)
			curveMaskValue = curveMaskDataHandle.asNurbsCurve()

			sculptedMeshDataHandle = _dataBlock.inputValue(SculptNodeClass.m_sculptedMesh)
			sculptedMeshValue = sculptedMeshDataHandle.asMesh()

			sculptStrengthDataHandle = _dataBlock.inputValue(SculptNodeClass.m_sculptStrength)
			sculptStrengthValue = sculptStrengthDataHandle.asFloat()

			curveOffsetDataHandle = _dataBlock.inputValue(SculptNodeClass.m_curveOffset)
			curveOffsetValue = curveOffsetDataHandle.asFloat()

			maxProjectionDistanceDataHandle = _dataBlock.inputValue(SculptNodeClass.m_maxProjectionDistance)
			maxProjectionDistanceValue = maxProjectionDistanceDataHandle.asFloat()

			outMeshDataHandle = _dataBlock.outputValue(SculptNodeClass.m_outMesh)

			# Computation
			# Create a copy of the terrain
			meshDataFn = om.MFnMeshData()
			outTerrain = meshDataFn.create()
			outTerrainFn = om.MFnMesh()
			outTerrainFn.copy(terrainValue, outTerrain)

			# Get all the vertices from the terrain
			inTerrainFn = om.MFnMesh(terrainValue)
			vertexPositions = inTerrainFn.getPoints()

			# Create a function set for the curve and find the centre
			curveFn = om.MFnNurbsCurve(curveMaskValue)
			curveCentre = self.findCurveCentre(curveFn)

			# If this is the first computation, store the curve positions as the original and compute the affectedVertices
			if self.m_curveOriginalPoints == None:
				self.m_curveOriginalPoints = curveFn.cvPositions(om.MSpace.kWorld)
				# Create a polygon iterator
				polygonIterator = om.MItMeshPolygon(terrainValue)
				# Find the closest face on the terrain
				centreFaceIndex = inTerrainFn.getClosestPoint(curveCentre, om.MSpace.kWorld)[1]
				# Calculate the affected vertices
				self.m_affectedVertices = self.findVerticesInsideCurve(polygonIterator, centreFaceIndex, curveFn, curveCentre, curveOffsetValue)
				# Store values
				self.m_lastCurveOffset = curveOffsetValue
				self.m_lastNumVertices = inTerrainFn.numVertices
				self.m_lastVertex01.append(inTerrainFn.getPoint(0, om.MSpace.kWorld))
				self.m_lastVertex01.append(inTerrainFn.getPoint(1, om.MSpace.kWorld))
			# Check if the input terrain, curve offset or the curve has changed
			else:
				recomputeAffectedVertices = False
				if self.m_lastNumVertices != inTerrainFn.numVertices:
					recomputeAffectedVertices = True
					self.m_lastNumVertices = inTerrainFn.numVertices
				elif curveOffsetValue != self.m_lastCurveOffset:
					recomputeAffectedVertices = True
					self.m_lastCurveOffset = curveOffsetValue
				elif self.m_lastVertex01[0] != inTerrainFn.getPoint(0, om.MSpace.kWorld) or self.m_lastVertex01[1] != inTerrainFn.getPoint(1, om.MSpace.kWorld):
					recomputeAffectedVertices = True
					self.m_lastVertex01[0] = inTerrainFn.getPoint(0, om.MSpace.kWorld)
					self.m_lastVertex01[1] = inTerrainFn.getPoint(1, om.MSpace.kWorld)
				elif len(self.m_curveOriginalPoints) != curveFn.numCVs:
					recomputeAffectedVertices = True
				else:
					curvePoints = curveFn.cvPositions(om.MSpace.kWorld)
					for originalPos, newPos in zip(self.m_curveOriginalPoints, curvePoints):
						if originalPos != newPos:
							recomputeAffectedVertices = True
							break
				if recomputeAffectedVertices == True:
					# Create a polygon iterator
					polygonIterator = om.MItMeshPolygon(terrainValue)
					# Find the closest face on the terrain
					centreFaceIndex = inTerrainFn.getClosestPoint(curveCentre, om.MSpace.kWorld)[1]
					self.findVerticesInsideCurve(polygonIterator, centreFaceIndex, curveFn, curveCentre, curveOffsetValue)
					self.m_curveOriginalPoints == curveFn.cvPositions(om.MSpace.kWorld)

			# Create a function set for the sculpted mesh
			sculptedMeshFn = om.MFnMesh(sculptedMeshValue)
			accelerationParams = sculptedMeshFn.autoUniformGridParams()

			# Iterate through affected vertices and project onto the sculpted mesh
			numVertices = len(self.m_affectedVertices)
			for i in xrange(numVertices):
				# Find a ray intersection from the original point in the direction of the normal to the scul mesh
				raySource = om.MFloatPoint(vertexPositions[self.m_affectedVertices[i]])
				normal = inTerrainFn.getVertexNormal(self.m_affectedVertices[i], True, om.MSpace.kWorld)
				intersection = sculptedMeshFn.closestIntersection(raySource, om.MFloatVector(normal), om.MSpace.kWorld, maxProjectionDistanceValue, False, accelParams=accelerationParams)
				# Calculate a vector from the original point to the new point
				difference = om.MPoint(intersection[0]) - vertexPositions[self.m_affectedVertices[i]]
				# Ensure the vertices are not sliding perpendicular to the normal
				if difference * normal != 0.0:
					closestPointOnCurve = curveFn.closestPoint(vertexPositions[self.m_affectedVertices[i]], space=om.MSpace.kWorld)[0]
					vertexPositions[self.m_affectedVertices[i]] += difference * sculptStrengthValue * self.calculateSoftSelectValue(curveCentre, vertexPositions[self.m_affectedVertices[i]], closestPointOnCurve)

			outTerrainFn.setPoints(vertexPositions)

			# Set the output value
			outMeshDataHandle.setMObject(outTerrain)

			# Mark the plug as clean
			outMeshDataHandle.setClean()

	## Calculate the soft selection value
	# @param _centre The centre of the curve
	# @param _vertex The vertex position to calculate for
	# @param _curvePoint The closest point on the curve to the vertex
	# @return A float value in the range [0,1]
	def calculateSoftSelectValue(self, _centre, _vertex, _curvePoint):
		centreVertexLenSquared = ((_centre.x - _vertex.x) * (_centre.x - _vertex.x)) + ((_centre.y - _vertex.y) * (_centre.y - _vertex.y)) + ((_centre.z - _vertex.z) * (_centre.z - _vertex.z))
		centreCurvePointLenSquared = ((_centre.x - _curvePoint.x) * (_centre.x - _curvePoint.x)) + ((_centre.y - _curvePoint.y) * (_centre.y - _curvePoint.y)) + ((_centre.z - _curvePoint.z) * (_centre.z - _curvePoint.z))
		ratio = centreVertexLenSquared / centreCurvePointLenSquared
		return 2 + 2.0/(ratio - 2)

	## Find the centre of a curve
	# @param _curveFn The curve function set
	# @return The centre of the curve as a MPoint
	def findCurveCentre(self, _curveFn):
		curvePoints = om.MPointArray()
		numPoints = _curveFn.numCVs * 2
		curvePoints.setLength(numPoints)
		curveCentre = om.MVector(0.0,0.0,0.0)
		for i in range(numPoints):
			point = _curveFn.getPointAtParam(float(i) / numPoints, om.MSpace.kWorld)
			curvePoints[i] = point
			curveCentre += om.MVector(point)
		# Find the centre of the curve and then typecast to MPoint
		curveCentre /= float(numPoints)
		return om.MPoint(curveCentre)

	## Find all the vertices inside the curve
	# @param _polygonIt The iterator for the mesh polygons
	# @param _startIndex The starting index for the iterator
	# @param _curveFn The curve function set
	# @param _curveCentre The centre of the curve
	# @param _curveOffset Offset the curve so it is still visible
	def findVerticesInsideCurve(self, _polygonIt, _startIndex, _curveFn, _curveCentre, _curveOffset):
		# Create a list and set containing face IDs
		uncheckedFaces = om.MIntArray() # This is a list of face IDs to check if they lie inside the curve
		checkedFaces = set() # This is a list of all the faces already checked, to avoid adding duplicates to uncheckedFaces
		# Add the starting index to this list and set
		uncheckedFaces.append(_startIndex)
		checkedFaces.add(_startIndex)

		# Create a list of vertex IDs that lie within the curve
		polyVertices = om.MIntArray()

		# Loop through and check the array of unchecked faces
		while (len(uncheckedFaces) > 0):
			# Set the iterator to the last item in the list of unchecked faces
			_polygonIt.setIndex(uncheckedFaces[-1])
			# Pop from the list of unchecked faces
			uncheckedFaces.remove(-1)
			# Find the centre of this face and the closest point on the curve
			faceCentre = _polygonIt.center(om.MSpace.kWorld)
			closestPointOnCurve = _curveFn.closestPoint(faceCentre, space=om.MSpace.kWorld)[0]
			# Calculate two vectors from the centre of the curve to the face and to the closest point on the curve
			centreToFace = (faceCentre - _curveCentre) * _curveOffset
			centreToCurve = closestPointOnCurve - _curveCentre
			# Use the dot product to see if the point is inside
			if centreToFace * centreToCurve.normal() < centreToCurve.length():
				# Add the vertices to the list of vertices that lie inside
				thisFaceVertices = _polygonIt.getVertices()
				polyVertices += thisFaceVertices
				# Find the connected faces and append to the list of unchecked faces
				connectedFaces = _polygonIt.getConnectedFaces()
				for faceId in connectedFaces:
					if faceId not in checkedFaces:
						uncheckedFaces.append(faceId)
						checkedFaces.add(faceId)
		self.m_affectedVertices = self.removeDuplicates(polyVertices)

	## Given a list, remove duplicate items
	# @param _list The list containing duplicates
	# @return A new list without any duplicates
	def removeDuplicates(self, _list):
		newList = []
		seenElements = set()
		for item in _list:
			if item not in seenElements:
				newList.append(item)
				seenElements.add(item)
		return newList

#----------------------------------------------------------
# Plugin Initialisation
#----------------------------------------------------------

## This function tells Maya to use the Python API 2.0
def maya_useNewAPI():
	pass

## Create an instance of the node
def nodeCreator():
	return SculptNodeClass()

## Initialise the node attributes
def nodeInitializer():
	# Create a numeric and typed attribute function set
	numericAttr = om.MFnNumericAttribute()
	typedAttr = om.MFnTypedAttribute()

	# Input node attributes
	SculptNodeClass.m_terrain = typedAttr.create("terrain", "t", om.MFnData.kMesh)
	typedAttr.readable = False
	typedAttr.writable = True
	typedAttr.storable = True
	SculptNodeClass.addAttribute(SculptNodeClass.m_terrain)

	SculptNodeClass.m_curveMask = typedAttr.create("curveMask", "cm", om.MFnData.kNurbsCurve)
	typedAttr.readable = False
	typedAttr.writable = True
	typedAttr.storable = True
	SculptNodeClass.addAttribute(SculptNodeClass.m_curveMask)

	SculptNodeClass.m_sculptedMesh = typedAttr.create("sculptedMesh", "sm", om.MFnData.kMesh)
	typedAttr.readable = False
	typedAttr.writable = True
	typedAttr.storable = True
	SculptNodeClass.addAttribute(SculptNodeClass.m_sculptedMesh)

	SculptNodeClass.m_sculptStrength = numericAttr.create("sculptStrength", "ss", om.MFnNumericData.kFloat, 1.0)
	numericAttr.readable = False
	numericAttr.writable = True
	numericAttr.storable = True
	SculptNodeClass.addAttribute(SculptNodeClass.m_sculptStrength)

	SculptNodeClass.m_curveOffset = numericAttr.create("curveOffset", "co", om.MFnNumericData.kFloat, 1.1)
	numericAttr.readable = False
	numericAttr.writable = True
	numericAttr.storable = True
	SculptNodeClass.addAttribute(SculptNodeClass.m_curveOffset)

	SculptNodeClass.m_maxProjectionDistance = numericAttr.create("maxProjectionDistance", "mpd", om.MFnNumericData.kFloat, 1000)
	numericAttr.readable = False
	numericAttr.writable = True
	numericAttr.storable = True
	SculptNodeClass.addAttribute(SculptNodeClass.m_maxProjectionDistance)

	# Output node attribute
	SculptNodeClass.m_outMesh = typedAttr.create("outMesh", "m", om.MFnData.kMesh)
	typedAttr.readable = True
	typedAttr.writable = False
	typedAttr.storable = False
	SculptNodeClass.addAttribute(SculptNodeClass.m_outMesh)

	# Connect input/output dependencies
	SculptNodeClass.attributeAffects(SculptNodeClass.m_terrain, SculptNodeClass.m_outMesh)
	SculptNodeClass.attributeAffects(SculptNodeClass.m_curveMask, SculptNodeClass.m_outMesh)
	SculptNodeClass.attributeAffects(SculptNodeClass.m_sculptedMesh, SculptNodeClass.m_outMesh)
	SculptNodeClass.attributeAffects(SculptNodeClass.m_sculptStrength, SculptNodeClass.m_outMesh)
	SculptNodeClass.attributeAffects(SculptNodeClass.m_curveOffset, SculptNodeClass.m_outMesh)
	SculptNodeClass.attributeAffects(SculptNodeClass.m_maxProjectionDistance, SculptNodeClass.m_outMesh)

## Initialise the plugin when Maya loads it
def initializePlugin(mobject):
	mplugin = om.MFnPlugin(mobject)
	try:
		mplugin.registerNode(kPluginNodeName, kPluginNodeID, nodeCreator, nodeInitializer)
	except:
		sys.stderr.write("Failed to register node: " + kPluginNodeName)
		raise

## Uninitialise the plugin when Maya unloads it
def uninitializePlugin(mobject):
	mplugin = om.MFnPlugin(mobject)
	try:
		mplugin.deregisterNode(kPluginNodeID)
	except:
		sys.stderr.write("Failed to unregister node: " + kPluginNodeName)
		raise
